pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -624,8 +624,8 @@ public UnAckedMessageTracker getUnAckedMessageTracker() {
                     if (message.hasKey()) {
                         typedMessageBuilderNew.key(message.getKey());
                     }
-                    typedMessageBuilderNew.send();
-                    return doAcknowledge(messageId, ackType, properties, null);
+                    return typedMessageBuilderNew.sendAsync().thenAccept(__ ->
```

Comment:
We should add `exceptionally()` section to handle the `CompletableFuture` failure and do negative-ack again on the message.

Reply:
+1. Can we collapse the logic for this conditional? Once both branches use `sendAsync`, they'll have a common future of type `CompletableFuture<MessageId>` and both call `doAcknowledge` in the callback. That will also collapse the exception handling logic.

Reply:
Yes, done.